projection/proofs/to_cake/payload_to_cakemlProofScript.sml

```diff
@@ -2822,29 +2838,84 @@ Theorem network_forward_correctness:
       (evaluate (st2 with clock := mc) env2
                       [compile_endpoint conf vs2 c'])
 Proof
-   rw []
-   \\ drule_then assume_tac NPar_trans_l_cases
-   \\ fs [] \\ rveq
-   >- (asm_exists_tac \\ fs []
-       \\ asm_exists_tac \\ fs []
-       \\ irule ffi_irrel \\ fs []
-       \\ conj_tac
-       >- metis_tac [ffi_wf_def,trans_ffi_eq_same]
-       \\ MAP_EVERY qexists_tac [‘p’,‘s’]
-       \\ rw [cpEval_valid_def,ffi_state_cor_def]
-       >- rw [ffi_wf_def]
-       \\ irule internal_trans_pres_wf
-       \\ MAP_EVERY qexists_tac [‘n’,‘conf’,‘s.queues’]
-       \\ rw [ffi_wf_def] \\ irule RTC_SINGLE
-       \\ rw [comms_ffi_consTheory.internal_trans_def]
-       \\ last_x_assum (assume_tac o ONCE_REWRITE_RULE [trans_cases]) \\ fs []
-       \\ IMP_RES_TAC trans_not_same \\ rw [] \\ fs [])
-   \\ drule_then (qspecl_then [‘vs1’,‘env1’,‘st1’] mp_tac) endpoint_forward_correctness
-   \\ CONV_TAC (PAT_CONV “λx. (∀vs2. x) ⇒ _” SWAP_FORALL_CONV)
-   \\ CONV_TAC (PAT_CONV “λx. x ⇒ _” SWAP_FORALL_CONV)
-   \\ disch_then (qspec_then ‘st2’ mp_tac)
-   \\ CONV_TAC (PAT_CONV “λx. x ⇒ _” SWAP_FORALL_CONV)
-   \\ cheat
+  rw []
+  \\ drule_then assume_tac NPar_trans_l_cases
+  \\ fs [] \\ rveq
+  (* p is not involved at all *)
+  >- (CONV_TAC SWAP_EXISTS_CONV
+      \\ asm_exists_tac \\ fs []
+      \\ asm_exists_tac \\ fs []
+      \\ irule ffi_irrel \\ fs []
+      \\ conj_tac
+      >- metis_tac [ffi_wf_def,trans_ffi_eq_same]
+      \\ MAP_EVERY qexists_tac [‘p’,‘s’]
+      \\ rw [cpEval_valid_def,ffi_state_cor_def]
+      >- rw [ffi_wf_def]
+      \\ irule internal_trans_pres_wf
+      \\ MAP_EVERY qexists_tac [‘n’,‘conf’,‘s.queues’]
+      \\ rw [ffi_wf_def] \\ irule RTC_SINGLE
+      \\ rw [comms_ffi_consTheory.internal_trans_def]
+      \\ last_x_assum (assume_tac o ONCE_REWRITE_RULE [trans_cases]) \\ fs []
+      \\ IMP_RES_TAC trans_not_same \\ rw [] \\ fs [])
+  (* LTau (only p does something) *)
+  >- (drule_then (qspecl_then [‘vs1’,‘env1’,‘st1’,‘st2’] mp_tac) endpoint_forward_correctness
+      \\ impl_tac
+      >- (rw [cpEval_valid_def,ffi_wf_def,
+              ffi_state_cor_def,
+              cpFFI_valid_def,some_def]
+          \\ fs []
+          \\ qpat_assum `trans _ (NEndpoint _ _ _) _ _`
+                          (mp_tac o PURE_ONCE_REWRITE_RULE [trans_cases])
+          \\ fs [] \\ rw []
+          \\ fs [state_component_equality] \\ rveq \\ rfs [state_component_equality]
+          >- (SELECT_ELIM_TAC \\ rw []
+              >- (asm_exists_tac \\ fs [])
+              \\ Cases_on ‘x'’ \\ fs []
+              \\ Cases_on ‘tl’ \\ rw [] >- cheat
+              \\ fs [normalise_queues_FUPDATE_NONEMPTY]
+              \\ cheat)
```

Comment:
This here is a `LTau` transition in payload that is trying to dequeue a message, but I'm having some trouble matching the transition in `payloadSem$trans` with the one in `strans`

Reply:
Is worth nothing that this is the case with `final d` in the assumptions

Reply:
This looks like a mismatch between how `trans` and `strans` handles queue normalisation. James moved away from association lists to finite maps to get us away from the `qcong` quagmire. I added queue normalisation to payload semantics to make sure the empty queue is always represented in the same way (otherwise, we would still need something akin to `qcong` to capture the fact that we identify `FEMPTY` and `FEMPTY |+ (q,[])` for the purposes of queue congruence. However, when I did this I didn't add queue normalisation to `strans`. From the looks of this goal state I think we should. My proposed fix would by to make the `ARecv` rule be like this:

```
  (∀conf c q ms mc tl N.
    qlk q ms = mc::tl
    ⇒ strans conf (c,q,N) (ARecv ms mc) (c,normalise_queues(q |+ (ms,tl)), N))
```

...which probably entails some minor repair work in James' confluence proofs.

After this, there should be no point case splitting on `tl`.

Reply:
Btw I'm taking a swing at making this change now